text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,182 @@
+- Feature Name: `layout_packed_aligned`
+- Start Date: 2024-10-24
+- RFC PR: [rust-lang/rfcs#3718](https://github.com/rust-lang/rfcs/pull/3718)
+- Rust Issue: [rust-lang/rust#100743](https://github.com/rust-lang/rust/issues/100743)
+
+# Summary
+[summary]: #summary
+
+This RFC makes it legal to have `#[repr(C)]` structs that are:
+- Both packed and aligned.
+- Packed, and transitively contains`#[repr(align)]` types.
+
+It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.
+It has the same behavior as `#[repr(C)]` except on `*-pc-windows-gnu` targets where it uses the msvc layout
+rules instead.
+
+# Motivation
+[motivation]: #motivation
+
+This RFC enables the following struct definitions:
+
+```rs
+#[repr(C, packed(2), align(4))]
+struct Foo { // Alignment = 4, Size = 8
+    a: u8,   // Offset = 0
+    b: u32,  // Offset = 2
+}
+```
+
+This is commonly needed when Rust is being used to interop with existing C and C++ code bases, which may contain
+unaligned types. For example in `clang` it is possible to create the following type definition, and there is
+currently no easy way to create a matching Rust type:
+
+```cpp
+struct  __attribute__((packed, aligned(4))) MyStruct {
+    uint8_t a;
+    uint32_t b;
+};
+```
+
+Currently, `#[repr(packed(_))]` structs cannot be `#[repr(align(_))]` or transitively contain `#[repr(align(_))]` types. Attempting to do so results in a [hard error](https://doc.rust-lang.org/nightly/error_codes/E0588.html).
+
+This behavior was added in the [original implementation](https://github.com/rust-lang/rust/issues/33158) of `#[repr(packed)]` due to concerns over differing behavior between msvc and gcc/clang. This makes it cumbersome or even impossible to produce C-compatible struct layouts in Rust when the corresponding C types were annotated with both `packed` and `aligned`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[repr(C)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
```

Comment:
Is there any difference in behavior between the c, system, and rust for types with both packed and align?

My expectation would be that regardless of whether it is c, system, or rust, if I have `repr(packed(2),align(4))`  the alignment of the overall type is 4, and the alignment of it's fields is at most 2 unless maybe one of those fields itself has an alignment specified. My reading of the reference section agrees with that, but the guide section is a little ambiguous. 

And FWIW, it would be more intuitive to me if packed ignored the alignment of field types, but had a way to specify higher alignment for individual fields.

Reply:
As currently written, `#[repr(align)]` used in a field's type definition (or transitively in its fields' types') could still raise the alignment of the type beyond the minimum alignment provided by the attribute, if the "infectious" alignment is in use (MSVC layout).

In MSVC C, such would theoretically be a compilation error (`alignas` cannot be used to lower alignment). I don't know what the behavior of templated C++ is, actually.

This is consistent with the behavior of `repr(align)` for an alignment less than the alignment required for primitive field alignment.